Sources/NIOHTTP2/HTTP2FrameParser.swift

```diff
@@ -0,0 +1,738 @@
+//
+//  HTTP2NativeParser.swift
+//  NIOHTTP2
+//
+//  Created by Jim Dovey on 11/14/18.
+//
+
+import NIO
+import NIOHPACK
+
+/// A temporary item used to hold onto the HPACK encoder/decoder. Something in the real stream state engine will
+/// take its place soon.
+class HTTP2StreamData {
+    static private var lookup: Dictionary<HTTP2StreamID, HTTP2StreamData> = [:]
+    
+    static func getStream(for streamID: HTTP2StreamID, allocator: ByteBufferAllocator) -> HTTP2StreamData {
+        if let stream = lookup[streamID] {
+            return stream
+        }
+        
+        let stream = HTTP2StreamData(streamID: streamID, allocator: allocator)
+        lookup[streamID] = stream
+        return stream
+    }
+    
+    var encoder: HPACKEncoder
+    var decoder: HPACKDecoder
+    
+    let streamID: HTTP2StreamID
+    
+    init(streamID: HTTP2StreamID, allocator: ByteBufferAllocator) {
+        // TODO: let's pass through a decent value for table sizes here.
+        self.encoder = HPACKEncoder(allocator: allocator)
+        self.decoder = HPACKDecoder(allocator: allocator)
+        self.streamID = streamID
+    }
+    
+    // internal so tests can see it.
+    class func reset() {
+        lookup.removeAll()
+    }
+}
+
+/// Ingests HTTP/2 data and produces frames. You feed data in, and sometimes you'll get a complete frame out.
+class HTTP2FrameDecoder {
+    private var streamDecoders: Dictionary<HTTP2StreamID, HPACKDecoder> = [:]
+    
+    private enum ParserState {
+        case idle
+        case accumulatingData(FrameHeader, ByteBuffer)
+    }
+    
+    private var state: ParserState = .idle
+    private var allocator: ByteBufferAllocator
+    private var currentHeaderFrame: HTTP2Frame? = nil
```

Comment:
My problem with that was that there's then a sort of sub-state issue here. We currently have 'idle' and 'collecting frame payload' states. Now there would be additional 'have some headers, but now idle' and 'have some headers, now collecting another frame's payload' states. As such, it seemed that the issue of there being a complete HEADERS frame sitting around waiting for continuations was outside the purview of the `ParserState` enum, since that relates only to a single frame's ingestion state.

Reply:
Actually, given that I need to accumulate all header fragments into a single buffer, this is now manageable — we'd have `idle`, `accumulating frame` and `accumulating header bytes` states.